lighthouse-core/scripts/i18n/collect-strings.js

```diff
@@ -382,82 +379,77 @@ const seenStrings = new Map();
-                if (!strings[seenId].meaning) {
-                  strings[seenId].meaning = strings[seenId].description;
-                  collisions++;
-                }
-                collisions++;
-              }
-            }
-
-            seenStrings.set(icuDefn.message, messageKey);
+function collectAllStringsInDir(dir) {
+  /** @type {Record<string, ICUMessageDefn>} */
+  const strings = {};
 
+  const globPattern = path.join(path.relative(LH_ROOT, dir), '/**/*.js');
+  const files = glob.sync(globPattern, {
+    cwd: LH_ROOT,
+    ignore: ignoredPathComponents,
+  });
+  for (const relativeToRootPath of files) {
+    const absolutePath = path.join(LH_ROOT, relativeToRootPath);
+    if (!process.env.CI) console.log('Collecting from', relativeToRootPath);
+
+    const content = fs.readFileSync(absolutePath, 'utf8');
+    const exportVars = require(absolutePath);
+    const regexMatches = UISTRINGS_REGEX.test(content);
+    const exportsUIStrings = Boolean(exportVars.UIStrings);
+    if (!regexMatches && !exportsUIStrings) continue;
+
+    if (regexMatches && !exportsUIStrings) {
+      throw new Error('UIStrings defined but not exported');
+    }
 
-            strings[messageKey] = icuDefn;
+    if (exportsUIStrings && !regexMatches) {
+      throw new Error('UIStrings exported but no definition found');
+    }
 
-            lastPropertyEndIndex = property.range[1];
+    // @ts-ignore regex just matched
+    const justUIStrings = 'const ' + content.match(UISTRINGS_REGEX)[0];
+    // just parse the UIStrings substring to avoid ES version issues, save time, etc
+    // @ts-ignore - esprima's type definition is supremely lacking
+    const ast = esprima.parse(justUIStrings, {comment: true, range: true});
+
+    for (const stmt of ast.body) {
+      if (stmt.type !== 'VariableDeclaration') continue;
+      if (stmt.declarations[0].id.name !== 'UIStrings') continue;
+
+      let lastPropertyEndIndex = 0;
+      for (const property of stmt.declarations[0].init.properties) {
+        const key = property.key.name;
+        const val = exportVars.UIStrings[key];
+        const {description, examples} = computeDescription(ast, property, val, lastPropertyEndIndex);
+        const converted = convertMessageToCtc(val, examples);
+        const messageKey = `${relativeToRootPath} | ${key}`;
+
+        /** @type {ICUMessageDefn} */
+        const icuDefn = {
+          message: converted.message,
+          description,
+          placeholders: converted.placeholders,
+        };
+
+        // check for duplicates, if duplicate, add @description as @meaning to both
+        if (seenStrings.has(icuDefn.message)) {
+          icuDefn.meaning = icuDefn.description;
+          const seenId = seenStrings.get(icuDefn.message);
+          if (seenId) {
+            if (!strings[seenId].meaning) {
+              strings[seenId].meaning = strings[seenId].description;
+              collisions++;
+            }
+            collisions++;
```

Comment:
+1 to @paulirish for keeping this global so this isn't a problem :smile:

Reply:
sorry i have no idea what your comment is referencing but this line of code looks like i screwed up a merge?

is it supposed to double count collisions

Reply:
nvm looks good